fusillade/clouddirectory.py

```diff
@@ -0,0 +1,788 @@
+import functools
+import json
+import typing
+import boto3
+from fusillade.errors import FusilladeException
+from collections import namedtuple
+from enum import Enum, auto
+from urllib.parse import quote, unquote
+import os
+
+ad = boto3.client("clouddirectory")
+arn = "arn:aws:clouddirectory:us-east-1:861229788715:"
+
+
+def get_published_schema_from_directory(dir_arn: str) -> str:
+    schema = ad.list_applied_schema_arns(DirectoryArn=dir_arn)['SchemaArns'][0]
+    schema = schema.split('/')[-2:]
+    schema = '/'.join(schema)
+    return f"{arn}schema/published/{schema}"
+
+
+def cleanup_directory(dir_arn: str):
+    ad.disable_directory(DirectoryArn=dir_arn)
+    ad.delete_directory(DirectoryArn=dir_arn)
+
+
+def cleanup_schema(sch_arn: str):
+    ad.delete_schema(SchemaArn=sch_arn)
+
+
+def publish_schema(name: str, version: str):
+    # don't create if already created
+    try:
+        dev_schema_arn = ad.create_schema(Name=name)['SchemaArn']
+    except ad.exceptions.SchemaAlreadyExistsException:
+        dev_schema_arn = f"{arn}schema/development/{name}"
+
+    # update the schema
+    pkg_root = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))  # noqa
+    with open(pkg_root + '/fusillade/directory_schema.json') as schema_json:
+        schema = json.dumps(json.load(schema_json))
+    ad.put_schema_from_json(SchemaArn=dev_schema_arn, Document=schema)
+    try:
+        pub_schema_arn = ad.publish_schema(DevelopmentSchemaArn=dev_schema_arn,
+                                           Version=version)['PublishedSchemaArn']
+    except ad.exceptions.SchemaAlreadyPublishedException:
+        pub_schema_arn = f"{arn}schema/published/{name}/{version}"
+    return pub_schema_arn
+
+
+def create_directory(name: str, schema: str):
+    try:
+        response = ad.create_directory(
+            Name=name,
+            SchemaArn=schema
+        )
+        directory = CloudDirectory(response['DirectoryArn'])
+
+        # create structure
+        directory.create_folder('/', 'Groups')
+        directory.create_folder('/', 'Users')
+        directory.create_folder('/', 'Roles')
+        directory.create_folder('/', 'Policies')
+
+        # create roles
+        with open("./policies/default_user_role.json", 'r') as fp:
+            Role.create(directory, "default_user", statement=fp.read())
+        with open("./policies/default_admin_role.json", 'r') as fp:
+            Role.create(directory, "admin", statement=fp.read())
+
+        # create admins
+        for admin in os.environ['FUS_ADMIN_EMAILS'].split(','):
+            user = User(directory, admin)
+            user.add_roles(['admin'])
+
+    except ad.exceptions.DirectoryAlreadyExistsException:
+        directory = CloudDirectory.from_name(name)
+    return directory
+
+
+def list_directories(state: str = 'ENABLED') -> typing.Iterator:
+    resp = ad.list_directories(state=state)
+    while True:
```

Comment:
This `while True` doesn't seem to be needed.  I'd keep it though and get rid of the recursion (which I feel is good whenever you can, especially since python has a restrictive stack of 999 recursions).  So something like:
```
def list_directories(state: str = 'ENABLED') -> typing.Iterator:
    while True:
        resp = ad.list_directories(state=state)
        for i in resp['Directories']:
            yield i
        if not resp.get('NextToken'):
            break
```

Reply:
this isn't a recursive function.

Reply:
Ah, it tripped me up because it has the same name as the function it's calling.  Maybe change the name to `list_dirs`?

Reply:
The shortened function should still work too.

Reply:
where would it be called twice?

Reply:
My bad, must be getting tired. Should be shortened though.